eth/filters/api.go

```diff
@@ -100,10 +98,11 @@ done:
 	for {
 		select {
 		case <-timer.C:
+			s.filterManager.Lock()
 			s.logMu.Lock()
 			for id, filter := range s.logQueue {
 				if time.Since(filter.timeout) > filterTickerTime {
-					s.filterManager.Remove(id)
+					s.filterManager.RemoveWhileLocked(id)
```

Comment:
Because of 4. in the original PR description. In the lines you quote, without the PR, logMu is acquired first, then Remove() tries to get sf.filterMu. While filterLoop() can concurrently first acquire sf.filterMu then through a callback (1) try to get logMu (2).

(1) https://github.com/ethereum/go-ethereum/blob/develop/eth/filters/filter_system.go#L158
(2) https://github.com/ethereum/go-ethereum/blob/develop/eth/filters/api.go#L225